vispy/visuals/markers.py

```diff
@@ -617,7 +618,7 @@ def set_data(self, pos=None, size=10., edge_width=1., edge_width_rel=None,
                 raise ValueError('edge_width_rel cannot be negative')
 
         if symbol is not None:
-            if not np.all(np.isin(np.asarray(symbol), self.symbols)):
+            if not all(x in self._symbol_shader_values for x in set(symbol if isinstance(symbol, Iterable) else [symbol])):
```

Comment:
I...this...this is faster for that 10^6 points case mentioned in the napari issue? When every symbol is valid (the common case) this has to check every symbol, right?

Another idea, what if we did `np.unique` on the passed in symbols and then checked for `isin` or contains?

Reply:
I don't want to distract, but for my own edification: doesn't the use of `set(symbol)` essentially do that?

Reply:
For a numpy array and for many elements I would assume numpy is faster.

Reply:
*assume* being the keyword here

Reply:
Wow this PR really is the fastest way:

```python
In [12]: symbols = np.random.randint(0, 7, size=10^6).astype(str)

In [13]: valid_symbols = ("1", "2", "3", "4", "5", "6")

In [14]: valid_symbols_dict = {x: x for x in valid_symbols}

In [15]: %%timeit
    ...: np.isin(np.unique(symbols), valid_symbols).all()
    ...:
    ...:
27.3 µs ± 2.16 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [16]: %%timeit
    ...: all(x in valid_symbols_dict for x in set(symbols))
    ...:
    ...:
4.64 µs ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [17]: %%timeit
    ...: all(x in valid_symbols_dict for x in np.unique(symbols))
    ...:
    ...:
6.27 µs ± 282 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
Ah I was missing a kwarg on the isin, but still way slower:

```
In [18]: %%timeit
    ...: np.isin(np.unique(symbols), valid_symbols, assume_unique=True).all()
    ...:
    ...:
25.6 µs ± 1.32 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
I should add this was Python 3.11.

Reply:
I also want to add, I wonder if this check can be combined with the line later down in the function where we convert to the GLSL integers:

```python
data['a_symbol'] = np.vectorize(self._symbol_shader_values.get)(symbol)
```

Overall I wonder if these two things could be done at the same time to save time.

Reply:
@djhoese your numbers seem very fast. I think your input may not be what you expect as `10^6` is only 12. You probably meant `10**6`.

Reply:
To your last point though - I do think this is faster. Benchmarking this vs. the proposed code I get about a 5x improvement. Here's what that might look like:

```diff
diff --git a/vispy/visuals/markers.py b/vispy/visuals/markers.py
index 0b8abb4d..73e1e539 100644
--- a/vispy/visuals/markers.py
+++ b/vispy/visuals/markers.py
@@ -617,9 +617,13 @@ class MarkersVisual(Visual):
             if np.any(edge_width_rel < 0):
                 raise ValueError('edge_width_rel cannot be negative')
 
+        symbols = []
         if symbol is not None:
-            if not all(x in self._symbol_shader_values for x in set(symbol if isinstance(symbol, Iterable) else [symbol])):
-                raise ValueError(f'symbols must one of {self.symbols}')
+            for x in set(symbol if isinstance(symbol, Iterable) and not isinstance(symbol, str) else [symbol]):
+                try:
+                    symbols.append(self._symbol_shader_values[x])
+                except KeyError:
+                    raise ValueError(f'symbols must be one of {self.symbols}')
 
         edge_color = ColorArray(edge_color).rgba
         if len(edge_color) == 1:
@@ -649,7 +653,7 @@ class MarkersVisual(Visual):
             data['a_position'][:, :pos.shape[1]] = pos
             data['a_size'] = size
 
-            data['a_symbol'] = np.vectorize(self._symbol_shader_values.get)(symbol)
+            data['a_symbol'] = symbols or None
 
             self._data = data
             self._vbo.set_data(data)
```

```python
import numpy as np
from vispy.visuals.markers import MarkersVisual
points = np.random.random((10_000_000, 2))
%timeit vis = MarkersVisual(pos=points, symbol=["o"] * 10**7)
```

```
# PR
1.53 s ± 6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# inlining conversion to int
309 ms ± 646 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
> For a numpy array and for many elements I would assume numpy is faster.

unfortunately not. current numpy implementation is naive. 

I have changed to set the difference. I do not know why it is faster as it should have a similar performance but looks faster.

Reply:
With set difference method on my above benchmark I get:
```
1.53 s ± 20.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
It may be faster for the first operation, but following @djhoese's suggestion to combine the conversion to integers is still much faster.

Reply:
Sorry for all the comments. I think this version is correct behavior and is still a little faster but not nearly as much:

```python
        symbols = []
        if symbol is not None:
            for x in [symbol] if isinstance(symbol, str) else symbol:
                try:
                    symbols.append(self._symbol_shader_values[x])
                except KeyError:
                    raise ValueError(f'symbols must be one of {self.symbols}')
```

Reply:
@aganders3 your code have 1s time of execution. Why?

Reply:
Ah sorry - this is for complete instantiation of the markers visual. Is yours much faster? I have not profiled to see what the next bottleneck is from here.
```python
import numpy as np
from vispy.visuals import MarkersVisual

points = np.random.random((10_000_000, 2))
%timeit vis = MarkersVisual(pos=points, symbol=["o"] * 10**7)
```

Reply:
> your numbers seem very fast. I think your input may not be what you expect as 10^6 is only 12. You probably meant 10**6.

lol that's great. Thanks for catching that.

Reply:
```python
In [19]: symbols = np.random.randint(0, 7, size=10**6).astype(str)

In [20]: valid_symbols = ("0", "1", "2", "3", "4", "5", "6")

In [21]: valid_symbols_dict = {x: idx for idx, x in enumerate(valid_symbols)}

In [22]: %%timeit
    ...: np.vectorize(valid_symbols_dict.get)(symbols)
    ...:
    ...:
80.4 ms ± 452 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Wrap it with a try/except TypeError and you don't need to do the explicit check:

```python
In [23]: valid_symbols = ("1", "2", "3", "4", "5", "6")

In [24]: valid_symbols_dict = {x: idx for idx, x in enumerate(valid_symbols)}

In [25]: %%timeit
    ...: np.vectorize(valid_symbols_dict.get)(symbols)
    ...:
    ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
...

TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'

```

Reply:
What I'm saying is, correct me if I'm wrong, we don't need to do the "is the user's provided symbol in our valid symbols". We just try to do the conversion and if it fails (exception) then we know why.

Reply:
The list comprehension is faster than numpy vectorize...

Reply:
Is it? What am I doing wrong?

```
In [31]: %%timeit
    ...: [valid_symbols_dict[x] for x in symbols]
    ...:
    ...:
226 ms ± 10.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [32]: %%timeit
    ...: np.vectorize(valid_symbols_dict.get)(symbols)
    ...:
    ...:
87 ms ± 4.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

10**6 values

Reply:
Even for 100 values it is still faster in numpy:

```
In [36]: symbols = np.random.randint(0, 7, size=10**2).astype(str)

In [37]:

In [37]: %%timeit
    ...: np.vectorize(valid_symbols_dict.get)(symbols)
    ...:
    ...:
16.5 µs ± 249 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [38]: %%timeit
    ...: [valid_symbols_dict[x] for x in symbols]
    ...:
    ...:
24.6 µs ± 1.3 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
This is my:


```python
In [6]: from vispy.visuals.markers import symbol_shader_values
In [7]: d = ["o" for _ in range(10**6)]

In [8]: %timeit np.vectorize(symbol_shader_values.get)(d)
195 ms ± 3.78 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: %timeit np.array([symbol_shader_values[x] for x in d])
59.8 ms ± 4.88 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

It may depend on the dict. 

Here another example:
```python
In [14]: d2 = [random.choice(list(symbol_shader_values)) for _ in range(10**6)]

In [15]: %timeit np.array([symbol_shader_values[x] for x in d2])
60.7 ms ± 1.07 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [16]: %timeit np.vectorize(symbol_shader_values.get)(d2)
244 ms ± 9.95 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Maybe I did something wrong?
(Python 3.11, but similar on python 3.9)

Reply:
@aganders3 Could you check on your machine?

Reply:
It depends if you're starting with a list or an array:

```python
In [44]: symbols = np.random.randint(0, 7, size=10**6).astype(str)

In [45]: symbols_list = symbols.tolist()

In [46]: %%timeit
    ...: [valid_symbols_dict[x] for x in symbols_list]
    ...:
    ...:
24.7 ms ± 618 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [47]: %%timeit
    ...: np.vectorize(valid_symbols_dict.get)(symbols_list)
    ...:
    ...:
166 ms ± 866 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [48]: %%timeit
    ...: [valid_symbols_dict[x] for x in symbols]
    ...:
    ...:
214 ms ± 4.66 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [49]: %%timeit
    ...: np.vectorize(valid_symbols_dict.get)(symbols)
    ...:
    ...:
81 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [50]: %%timeit
    ...: np.asarray(symbols_list)
    ...:
    ...:
96.7 ms ± 491 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [51]: %%timeit
    ...: np.asarray([valid_symbols_dict[x] for x in symbols])
    ...:
    ...:
255 ms ± 1.62 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [52]: %%timeit
    ...: np.asarray([valid_symbols_dict[x] for x in symbols_list])
    ...:
    ...:
62.6 ms ± 590 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The conversion from list to numpy array alone was 96ms on my machine. Python 3.11, numpy 1.24.4.

Reply:
Looks like `np.fromiter` is technically faster by a little bit too when starting from a list, but not by much.

```python
In [7]: %%timeit
   ...: np.fromiter((valid_symbols_dict[x] for x in symbols_list), np.uint32, count=len(symbols_list))
   ...:
   ...:
64.7 ms ± 956 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [8]: %%timeit
   ...: np.asarray([valid_symbols_dict[x] for x in symbols_list])
   ...:
   ...:
67.5 ms ± 1.75 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
```python
In [1]: import numpy as np

In [2]: import random

In [3]: from vispy.visuals.markers import symbol_shader_values

In [4]: d = ["o" for _ in range(10**6)]

In [5]: d_n = np.array(d, dtype=object)

In [6]: %timeit [symbol_shader_values[x] for x in d]
25.8 ms ± 1.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [7]: %timeit np.array([symbol_shader_values[x] for x in d])
55.3 ms ± 857 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [8]: %timeit np.vectorize(symbol_shader_values.get)(d)
196 ms ± 4.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit [symbol_shader_values[x] for x in d_n]
31.2 ms ± 1.17 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [10]: %timeit np.array([symbol_shader_values[x] for x in d_n])
67.9 ms ± 2.92 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [11]: %timeit np.vectorize(symbol_shader_values.get)(d_n)
64.1 ms ± 1.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [12]: d2 = [random.choice(list(symbol_shader_values)) for _ in range(10**6)]

In [13]: d2_n = np.array(d2, dtype=object)

In [14]: %timeit [symbol_shader_values[x] for x in d2]
25.2 ms ± 551 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [15]: %timeit np.array([symbol_shader_values[x] for x in d2])
63.5 ms ± 3.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [16]: %timeit np.vectorize(symbol_shader_values.get)(d2)
245 ms ± 18.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [17]: %timeit [symbol_shader_values[x] for x in d2_n]
32.8 ms ± 1.79 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [18]: %timeit np.array([symbol_shader_values[x] for x in d2_n])
66.5 ms ± 807 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [19]: %timeit np.vectorize(symbol_shader_values.get)(d2_n)
68.6 ms ± 3.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

This is my linux machine (Ubuntu 22.04, python 3.11.5, numpy 1.26.0)

Reply:
Do your arrays have to be of type object? Why not a string type?

Reply:
Because it save time for conversion back to python string?

```
In [29]: d2_n = np.array(d2, dtype=object)

In [30]: %timeit np.vectorize(symbol_shader_values.get)(d2_n)
68.2 ms ± 2.24 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [31]: d2_n = np.array(d2, dtype=str)

In [32]: %timeit np.vectorize(symbol_shader_values.get)(d2_n)
175 ms ± 2.46 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
Ok, I did it with the actual dictionary from vispy (well copied and edited to just be a raw dictionary so I didn't have to import vispy) and `np.vectorize` performs much slower with more options from this bigger dictionary. The numpy array input is still faster for the vectorize function than passing it a list. This still means that numpy input works better with vectorize (~154ms on my machine) while a list input works better with the listcomp (~61ms). Very strange.

I also noticed that in both of the faster cases (for the given input) timeit is running 10 loops each but 1 loop each for the slower methods. We must be right on the edge of a threshold.

Reply:
> Because it save time for conversion back to python string?

Haha I hadn't even realized that. Mine went from ~152ms to ~47ms.

Reply:
So, the big question is how the user will provide symbols. I more expect Python list, but I may not have a proper overview of common vispy usecases.

Reply:
I'm not sure I can say what is more common here. This features isn't that old actually and was added by @brisvag in https://github.com/vispy/vispy/pull/2361. So napari might be the leader on this so you'll have to tell me what you want to do. Theoretically if these are numbers we'd probably want to do everything in numpy to save on the memory and make access faster and lead to the potential of good vectorization. That said, we're dealing with strings and generating a list of string combinations or combining lists of strings is a lot easier that trying to concatenate numpy arrays of strings. At least in my opinion.

The docstring does only say str or array though. :man_shrugging:

Reply:
It also is on the table on our side as we now use `StrEnum` that provides big penalty on comparission. 
https://github.com/napari/napari/pull/6277

Reply:
Great read, nice invertigation! I would asssume that in the only cases where speed really matters (= big number of points), people are more likely to be using numpy array.